drivers: serial: binding: add devicetree init support

JvanDooren opened the pull request:
Extend the binding to allow buildtime configuration of
- stopbits
- databits

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

JvanDooren committed:
drivers: serial: binding: add devicetree init support

Extend the binding to allow compiletime configuration of
of the stopbits and databits.

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

gmarull commented on dts/bindings/serial/uart-controller.yaml:
```diff
@@ -23,3 +23,22 @@ properties:
+  stop-bits:
+    type: string
+    description: |
+      Sets the number of stop bits. Default to 1 if not specified.
+    enum:
+      - "0_5"
+      - "1"
+      - "1_5"
+      - "2"
+  data-bits:
+    type: string
+    description: |
+      Sets the number of data bits. Default to 8 if not specified.
```
- this doesn't justify defaults (required).
- seems to deviate from https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/serial/serial.yaml cc: @galak 
- how is this going to be used? app needs to call uart_configure() where those are passed.

nordicjm replied:
> how is this going to be used? app needs to call uart_configure() where those are passed.

Through normal driver initialisation? E.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_rpi_pico.c#L145 instead of these being static values, they can be retrieved from the devicetree for a UART instance.

gmarull replied:
If we allow initial config from DT, then I guess we need to use the `zephyr,` prefix here. Then, this is something that should be handled by the driver class itself, to avoid repeating the same code on every single driver. See e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/regulator/regulator_common.c#L16 (still not ideal, our device model isn't great to do these things nicely).

JvanDooren replied:
I understand it doesn't justify defaults, because if there is no implementation behind it (like for none of the drivers), then they will require runtime configuration? There probably is no good default to choose because it depends on the arch and platform? Then why is there a mention of default on parity?

heinwessels replied:
Removed mention of defaults.

galak reviewed (changes requested):
I don't have any issues w/this, but would like to see if there's actually some driver updates to utilize this.

Would like @dcpleung ack on this as well.

JvanDooren commented:
@galak @dcpleung  Showcase is here: https://github.com/zephyrproject-rtos/zephyr/pull/54354

dcpleung commented:
> @galak @dcpleung Showcase is here: #54354

In that case, move this commit to #54354 and place it before the UART driver commits. So that these can be reviewed together.

JvanDooren commented:
> 

This is already as first commit there, but I was told to split it off.

dcpleung commented:
> This is already as first commit there, but I was told to split it off.

Ahh... in that case... I think removing the "default" wording would be great. Or else you will need to make sure every driver use those as defaults.

Though, I still think it is a good idea to implement a helper function to grab the device tree settings and call `uart_configuate()`.

JvanDooren commented:
> 

1. About the 'default' wording:
Ok, I understand your point. However, in my defense, the current binding also contains a parity 'Default', which is indeed the default in the serial drivers I looked at.
Most drivers don't even use the binding and have hardcoded .uart_config.parity = UART_CFG_PARITY_NONE.

Before my change, all drivers using devicetree initialization used a default internally for databits+stopbits (which coincidentally appears to be 8N1) without any mention in any binding. So, apparently we rolled the dice, gambled that it was 8N1 and were lucky. Is my word 'Default' so wrong then?

2. About the helper function:
Yes, that would be superb, to have such a function. I however don't see how this is going to work out, because we're talking about using the DT_INST_ENUM_IDX_OR(index, data-bits, UART_CFG_PARITY_NONE) inside a architecture driver specific structure such as

```
static struct uart_ns16550_dev_data uart_ns16550_dev_data_##n = {            \
		.uart_config.baudrate = DT_INST_PROP_OR(n, current_speed, 0),        \
		.uart_config.parity = UART_CFG_PARITY_NONE,                          \
		.uart_config.stop_bits = UART_CFG_STOP_BITS_1,                       \
		.uart_config.data_bits = UART_CFG_DATA_BITS_8,                       \
		.uart_config.flow_ctrl = DEV_DATA_FLOW_CTRL(n),                      \
		DEV_DATA_DLF_INIT(n)                                                 \
	};
```
Each driver has its own structure, wrapping or even replacing the struct uart_config so how would that work? In every devicetree-init method, we'd have to call this generic method with a 'struct uart_config' to fill in the parity/start/stop/data-bits. Then it's up to every driver to use these or copy these right?
		\
```

dcpleung commented:
* Remove the "Default" wording for now since not all drivers are using those defaults.
* You can provide an helper initializer to `struct uart_config` using the device tree properties. That way, each driver can feed it into their own configuration function.

galak commented on dts/bindings/serial/uart-controller.yaml:
```diff
@@ -23,3 +23,22 @@ properties:
+      - "1_5"
+      - "2"
+  data-bits:
+    type: string
```
Why is this a string?

heinwessels replied:
I guess it was to be consistent with `stop-bits`, but it doesn't have to be. Changed to `int`. Also did a rebase.

heinwessels committed:
dts: bindings: serial: infineon cat1-uart: remove properties

These bindings are removed because they are now handled in the
included base uart-controller.yml.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

ifyall reviewed (approved):
Approving for the change to the infineon.cat1-uart.yaml.
